sql/pgwire: TestAuthenticationAndHBARules failed

[(sql/pgwire).TestAuthenticationAndHBARules failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1992368&tab=buildLog) on [master@dfb6e1bee4abc2f2ff82d545686908d983f82872](https://github.com/cockroachdb/cockroach/commits/dfb6e1bee4abc2f2ff82d545686908d983f82872):

```
=== RUN   TestAuthenticationAndHBARules/insecure=false/insecure
I200606 03:21:05.504564 35436 base/addr_validation.go:296  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I200606 03:21:05.504727 35436 base/addr_validation.go:342  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
W200606 03:21:05.505618 35436 server/status/runtime.go:309  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I200606 03:21:05.573149 35436 server/server.go:748  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I200606 03:21:05.577501 35436 server/config.go:595  [n?] 1 storage engine initialized
I200606 03:21:05.577757 35436 server/config.go:598  [n?] Pebble cache size: 128 MiB
I200606 03:21:05.577883 35436 server/config.go:598  [n?] store 0: in-memory, size 0 B
I200606 03:21:05.633348 35436 server/server.go:1245  [n?] **** add additional nodes by specifying --join=127.0.0.1:40927
I200606 03:21:05.633828 35436 server/init.go:173  [n?] no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.
I200606 03:21:05.634017 35436 server/init.go:193  [n?] **** cluster 76e747f8-0fa8-41a3-a6b6-6b201e1aac61 has been created
I200606 03:21:05.634804 35436 gossip/gossip.go:405  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:40927" > attrs:<> locality:<> ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:7 > build_tag:"v20.2.0-alpha.00000000-2059-gdfb6e1bee4" started_at:1591413665634480395 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.0.0.1:36697" > 
I200606 03:21:05.642374 35408 server/server.go:1327  [n1] connecting to gossip network to verify cluster ID "76e747f8-0fa8-41a3-a6b6-6b201e1aac61"
I200606 03:21:05.678164 35408 server/server.go:1330  [n1] node connected via gossip
W200606 03:21:05.678336 35611 kv/kvserver/store.go:1652  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200606 03:21:05.688124 35436 server/node.go:412  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=18 KiB), ranges=34, leases=2, queries=0.00, writes=0.00, bytesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=166.00 pMax=15513.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I200606 03:21:05.688562 35436 kv/kvserver/stores.go:236  [n1] read 0 node addresses from persistent storage
I200606 03:21:05.689333 35436 server/node.go:468  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200606 03:21:05.716844 35436 server/server.go:1426  [n1] starting https server at 127.0.0.1:46029 (use: 127.0.0.1:46029)
I200606 03:21:05.717341 35436 server/server.go:1431  [n1] starting postgres server at 127.0.0.1:36697 (use: 127.0.0.1:36697)
I200606 03:21:05.717409 35436 server/server.go:1433  [n1] starting grpc server at 127.0.0.1:40927
I200606 03:21:05.717459 35436 server/server.go:1434  [n1] advertising CockroachDB node at 127.0.0.1:40927
I200606 03:21:05.806247 35752 sql/temporary_schema.go:473  [n1] running temporary object cleanup background job
I200606 03:21:05.828962 35680 kv/kvserver/replica_rangefeed.go:608  [n1,s1,r6/1:/Table/{SystemCon…-11}] RangeFeed closed timestamp is empty
I200606 03:21:05.848825 35752 sql/temporary_schema.go:508  [n1] found 0 temporary schemas
I200606 03:21:05.849423 35752 sql/temporary_schema.go:511  [n1] early exiting temporary schema cleaner as no temporary schemas were found
I200606 03:21:05.849828 35752 sql/temporary_schema.go:512  [n1] completed temporary object cleanup job
I200606 03:21:05.850217 35752 sql/temporary_schema.go:595  [n1] temporary object cleaner next scheduled to run at 2020-06-06 03:51:05.805881316 +0000 UTC
I200606 03:21:05.859794 35751 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.0.0.1:40927 Attrs: Locality: ServerVersion:20.1-7 BuildTag:v20.2.0-alpha.00000000-2059-gdfb6e1bee4 StartedAt:1591413665634480395 LocalityAddress:[] ClusterName: SQLAddress:127.0.0.1:36697} ClusterID:76e747f8-0fa8-41a3-a6b6-6b201e1aac61 StartedAt:1591413665634480395 LastUp:1591413665634480395}
I200606 03:21:05.882088 35633 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200606 03:21:05.917154 35750 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200606 03:21:05.962378 35799 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1-7 User:root}
I200606 03:21:06.099204 36085 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:0d385601-282f-437c-aa29-425d9799b420 User:root}
I200606 03:21:06.140545 36199 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200606 03:21:06.168884 36200 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200606 03:21:06.289100 35436 server/server_sql.go:645  [n1] done ensuring all necessary migrations have run
I200606 03:21:06.289208 35436 server/server.go:1747  [n1] serving sql connections
I200606 03:21:06.289414 35436 server/server.go:1768  [n1] starting postgres server at unix:/tmp/TestAuth329856130/.s.PGSQL.6
I200606 03:21:06.290329 36123 sqlmigrations/migrations.go:727  [n1] starting wait for upgrade finalization before schema change job migration
I200606 03:21:06.294764 36123 sqlmigrations/migrations.go:759  [n1] starting schema change job migration
FAIL	github.com/cockroachdb/cockroach/pkg/sql/pgwire	109.459s

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestAuthenticationAndHBARules PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAuthenticationAndHBARules.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/pgwire).TestAuthenticationAndHBARules failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1993050&tab=buildLog) on [master@dfb6e1bee4abc2f2ff82d545686908d983f82872](https://github.com/cockroachdb/cockroach/commits/dfb6e1bee4abc2f2ff82d545686908d983f82872):

```
I200606 06:30:27.534422 47556 server/config.go:595  [n?] 1 storage engine initialized
I200606 06:30:27.534516 47556 server/config.go:598  [n?] Pebble cache size: 128 MiB
I200606 06:30:27.534559 47556 server/config.go:598  [n?] store 0: in-memory, size 0 B
I200606 06:30:27.576665 47556 server/server.go:1245  [n?] **** add additional nodes by specifying --join=127.113.82.129:38575
I200606 06:30:27.576927 47556 server/init.go:173  [n?] no stores bootstrapped and --join flag specified, awaiting init command or join with an already initialized node.
I200606 06:30:27.577016 47556 server/init.go:193  [n?] **** cluster fc79cf7a-9f62-4334-96f2-072f83f3d42c has been created
I200606 06:30:27.577465 47790 server/server.go:1327  [n1] connecting to gossip network to verify cluster ID "fc79cf7a-9f62-4334-96f2-072f83f3d42c"
I200606 06:30:27.577601 47556 gossip/gossip.go:405  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.113.82.129:38575" > attrs:<> locality:<> ServerVersion:<major_val:20 minor_val:1 patch:0 unstable:7 > build_tag:"v20.2.0-alpha.00000000-2059-gdfb6e1bee4" started_at:1591425027577341483 cluster_name:"" sql_address:<network_field:"tcp" address_field:"127.113.82.129:41715" > 
I200606 06:30:27.615760 47790 server/server.go:1330  [n1] node connected via gossip
W200606 06:30:27.616063 47869 kv/kvserver/store.go:1652  [n1,s1,r6/1:/Table/{SystemCon…-11}] could not gossip system config: [NotLeaseHolderError] r6: replica (n1,s1):1 not lease holder; lease holder unknown
I200606 06:30:27.619841 47556 server/node.go:412  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=18 KiB), ranges=34, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=166.00 pMax=15513.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I200606 06:30:27.620325 47556 kv/kvserver/stores.go:236  [n1] read 0 node addresses from persistent storage
I200606 06:30:27.621076 47556 server/node.go:468  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I200606 06:30:27.622680 47556 server/server.go:1426  [n1] starting http server at 127.113.82.129:44541 (use: 127.113.82.129:44541)
I200606 06:30:27.622788 47556 server/server.go:1431  [n1] starting postgres server at 127.113.82.129:41715 (use: 127.113.82.129:41715)
I200606 06:30:27.622880 47556 server/server.go:1433  [n1] starting grpc server at 127.113.82.129:38575
I200606 06:30:27.622944 47556 server/server.go:1434  [n1] advertising CockroachDB node at 127.113.82.129:38575
I200606 06:30:27.713188 47963 sql/temporary_schema.go:473  [n1] running temporary object cleanup background job
I200606 06:30:27.725981 48034 kv/kvserver/replica_rangefeed.go:608  [n1,s1,r6/1:/Table/{SystemCon…-11}] RangeFeed closed timestamp is empty
I200606 06:30:27.737089 47963 sql/temporary_schema.go:508  [n1] found 0 temporary schemas
I200606 06:30:27.737218 47963 sql/temporary_schema.go:511  [n1] early exiting temporary schema cleaner as no temporary schemas were found
I200606 06:30:27.737276 47963 sql/temporary_schema.go:512  [n1] completed temporary object cleanup job
I200606 06:30:27.737386 47963 sql/temporary_schema.go:595  [n1] temporary object cleaner next scheduled to run at 2020-06-06 07:00:27.71277277 +0000 UTC
I200606 06:30:27.751251 47962 sql/event_log.go:132  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:127.113.82.129:38575 Attrs: Locality: ServerVersion:20.1-7 BuildTag:v20.2.0-alpha.00000000-2059-gdfb6e1bee4 StartedAt:1591425027577341483 LocalityAddress:[] ClusterName: SQLAddress:127.113.82.129:41715} ClusterID:fc79cf7a-9f62-4334-96f2-072f83f3d42c StartedAt:1591425027577341483 LastUp:1591425027577341483}
I200606 06:30:27.775856 47912 sql/event_log.go:132  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I200606 06:30:27.811460 47961 server/server_update.go:55  [n1] no need to upgrade, cluster already at the newest version
I200606 06:30:27.846988 48111 sql/event_log.go:132  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:20.1-7 User:root}
I200606 06:30:27.982958 48302 sql/event_log.go:132  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:031619e2-4e28-4bf4-925f-96b250a7dbbd User:root}
I200606 06:30:28.037707 48531 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I200606 06:30:28.074192 48535 sql/event_log.go:132  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I200606 06:30:28.255851 47556 server/server_sql.go:645  [n1] done ensuring all necessary migrations have run
I200606 06:30:28.256007 47556 server/server.go:1747  [n1] serving sql connections
I200606 06:30:28.256275 47556 server/server.go:1768  [n1] starting postgres server at unix:/tmp/TestAuth553028591/.s.PGSQL.6
I200606 06:30:28.262142 48682 sqlmigrations/migrations.go:727  [n1] starting wait for upgrade finalization before schema change job migration
I200606 06:30:28.264328 48607 sql/pgwire/server.go:493  [n1,client=127.0.0.1:43048] 1 received connection
I200606 06:30:28.280282 48698 sql/pgwire/auth.go:335  [n1,client=127.0.0.1:43048,hostnossl,user=root] 2 connection matches HBA rule: 
I200606 06:30:28.282439 48698 sql/pgwire/auth.go:335  [n1,client=127.0.0.1:43048,hostnossl,user=root] 3 authentication succeeded
test logs left over in: /tmp/logTestAuthenticationAndHBARules_insecure=true_insecure733942419
I200606 06:30:28.334501 47556 util/stop/stopper.go:539  quiescing
I200606 06:30:28.334920 47867 kv/kvserver/queue.go:580  [n1,s1] rate limited in gossip update (merge): node unavailable; try another peer
W200606 06:30:28.335008 48537 kv/kvserver/intentresolver/intent_resolver.go:744  failed to gc transaction record: could not GC completed transaction anchored at /Table/SystemConfigSpan/Start: node unavailable; try another peer
I200606 06:30:28.335211 47923 kv/kvclient/kvcoord/transport_race.go:108  transport race promotion: ran 17 iterations on up to 271 requests
I200606 06:30:28.335351 47867 kv/kvserver/queue.go:580  [n1,s1] rate limited in gossip update (split): node unavailable; try another peer
I200606 06:30:28.342473 48607 sql/pgwire/conn.go:226  [n1,client=127.0.0.1:43048,hostnossl,user=root] 4 session terminated; duration: 62.416355ms
I200606 06:30:28.342665 48607 sql/pgwire/server.go:501  [n1,client=127.0.0.1:43048,hostnossl] 5 disconnected; duration: 78.356282ms
        --- FAIL: TestAuthenticationAndHBARules/insecure=true/insecure (0.87s)
            test_log_scope.go:77: test logs captured to: /tmp/logTestAuthenticationAndHBARules_insecure=true_insecure733942419
            testing.go:853: race detected during execution of test
=== RUN   TestAuthenticationAndHBARules/insecure=true
    --- FAIL: TestAuthenticationAndHBARules/insecure=true (9.87s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestAuthenticationAndHBARules PKG=./pkg/sql/pgwire TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAuthenticationAndHBARules.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/util/log/stderr_redirect.go
+++ b/pkg/util/log/stderr_redirect.go
@@ -10,7 +10,11 @@
 
 package log
 
-import "os"
+import (
+	"os"
+
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+)
 
 // OrigStderr points to the original stderr stream.
 var OrigStderr = func() *os.File {
@@ -44,3 +48,7 @@ func hijackStderr(f *os.File) error {
 func restoreStderr() error {
 	return redirectStderr(OrigStderr)
 }
+
+// osStderrMu ensures that concurrent redirects of stderr don't
+// overwrite each other.
+var osStderrMu syncutil.Mutex

--- a/pkg/util/log/stderr_redirect_unix.go
+++ b/pkg/util/log/stderr_redirect_unix.go
@@ -47,6 +47,8 @@ func dupFD(fd uintptr) (uintptr, error) {
 // We also override os.Stderr for those other parts of Go which use
 // that and not fd 2 directly.
 func redirectStderr(f *os.File) error {
+	osStderrMu.Lock()
+	defer osStderrMu.Unlock()
 	if err := unix.Dup2(int(f.Fd()), unix.Stderr); err != nil {
 		return err
 	}

--- a/pkg/util/log/stderr_redirect_windows.go
+++ b/pkg/util/log/stderr_redirect_windows.go
@@ -35,6 +35,8 @@ func dupFD(fd uintptr) (uintptr, error) {
 // We also override os.Stderr for those other parts of Go which use
 // that and not fd 2 directly.
 func redirectStderr(f *os.File) error {
+	osStderrMu.Lock()
+	defer osStderrMu.Unlock()
 	if err := windows.SetStdHandle(windows.STD_ERROR_HANDLE, windows.Handle(f.Fd())); err != nil {
 		return err
 	}

